interfaces/apparmor/apparmor.go

```diff
@@ -117,9 +118,31 @@ func loadProfiles(fnames []string, cacheDir string, flags aaParserFlags) error {
 	if !osutil.GetenvBool("SNAPD_DEBUG") {
 		args = append(args, "--quiet")
 	}
-	args = append(args, fnames...)
 
-	output, err := exec.Command("apparmor_parser", args...).CombinedOutput()
+	parser, internal, err := apparmor_sandbox.FindAppArmorParser()
+	if err != nil || !internal {
+		// if we couldn't find the parser with apparmor_sandbox
+		// then fall-back to trying to find one in the current PATH
+		// - same for if we are not using the internal
+		// apparmor_parser so that we can support a mocked parser
+		// during tests
+		parser = "apparmor_parser"
+	} else {
+		// when using the internal apparmor_parser also use it's
+		// own configuration and includes etc plus also ensure we
+		// use the 3.0 feature ABI to get the widest array of
+		// policy features across the widest array of kernels
+		// versions
+		prefix := strings.TrimSuffix(parser, "apparmor_parser")
+		args = append(args, []string{
+			"--config-file", filepath.Join(prefix, "/apparmor/parser.conf"),
+			"--base", filepath.Join(prefix, "/apparmor.d"),
+			"--policy-features", filepath.Join(prefix, "/apparmor.d/abi/3.0"),
+		}...)
+	}
+
+	args = append(args, fnames...)
+	output, err := exec.Command(parser, args...).CombinedOutput()
```

Comment:
it feels like maybe this code belongs to a sandbox/apparmor.RunParser helper?

Reply:
I thought perhaps having the sandbox specify any required arguments would be a more flexible approach - see https://github.com/alexmurray/snapd/commit/9411f94271c6fac36f2580440431f82b1447bc0d - but if you would prefer to have the sandbox do the exec via a helper etc I'm easy either way - if we did add a `RunParser()` method, would you want that to also encapsulate the other parameters which we currently have in interfaces/apparmor like `no-expr-simplify` etc - https://github.com/snapcore/snapd/blob/062ef6a1ace0456ff21c3e89ea524dd326ac7db6/interfaces/apparmor/apparmor.go#L102 ?

Reply:
Regarding the `#include if exists` - that is a good idea but since this is in the base-template it would get included in all profiles - if we are going to take this approach, would we also want to take the opportunity to allow per-profile includes as well? ie we could append to the bottom of the snap profile `#include if exists "/var/lib/snapd/apparmor/profiles/local/snap.SNAP_NAME.APP"` ?

The other option is to do what I suggested earlier and have snapd copy out the now vendored apparmor config etc to `/var/lib/snapd/apparmor/` so that it can be modified directly (but then I wonder how we would handle the case where we want to upgrade this config in some future where we update to say apparmor-3.1 when it comes along...)

Reply:
> Regarding the #include if exists - that is a good idea but since this is in the base-template it would get included in all profiles 

that's the intention, we would place the include such that it makes it easy to override things that come from the tunables

> - if we are going to take this approach, would we also want to take the opportunity to allow per-profile includes as well? ie we could append to the bottom of the snap profile #include if exists "/var/lib/snapd/apparmor/profiles/local/snap.SNAP_NAME.APP"

it seems two slightly unrelated problems. I know we ask people to play with per-snap profiles to test things, but this instead seems very upfront and maybe prone to bad habits

Reply:
The other issue with `include if exists` is that it is only supported by apparmor >= 2.13 - so this would not work on bionic or older which ship apparmor 2.12. As such I think we would only want to generate this when using the vendored apparmor but then in that case I think we should not refer to `/etc/apparmor.d/` but instead the snapd specific path of `/var/lib/snapd/apparmor/snap-tunables` or similar.

Reply:
To avoid any issues, I have added support for probing appamor_parser to see if it supports `include if exists` (https://github.com/snapcore/snapd/pull/9939/commits/02e0dc62246fdc86b108968d86739ca8b40f11cb) and only generating this rule if that is the case (https://github.com/snapcore/snapd/pull/9939/commits/9b2a2866cc6024a11887011a21e435b95be84ce1).

@pedronis + @mvo5 - let me know if there is anything else I can do to get this merged.. :crossed_fingers: